File: infinitic-common/src/test/java/io/infinitic/serDe/java/SerDeTests.java
package io.infinitic.serDe.java;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import io.infinitic.common.serDe.SerializedData;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;

class SerDeTest {

    @Test
    public void nullObjShouldBeSerializableDeserializable() {
        SerializedData data = SerializedData.from(null, null);

        Assertions.assertNull(data.deserialize(null));
    }

    @Test
    public void simpleObjectShouldBeSerializableDeserializable() {
        Pojo1 val1 = new Pojo1("42", 42, JType.TYPE_1);
        SerializedData data = SerializedData.from(val1, null);
        System.out.println(data.toJsonString());

        Assertions.assertEquals(val1, data.deserialize(null));
    }

    @Test
    public void objectShouldBeDeserializableEvenWithMoreProperties() {
        Pojo1 val1 = new Pojo1("42", 42, JType.TYPE_1);
        Pojo2 val2 = new Pojo2("42", 42);
        HashMap<String, byte[]> meta = new HashMap<>();
        meta.put(SerializedData.META_JAVA_CLASS, Pojo2.class.getName().getBytes(StandardCharsets.UTF_8));
        SerializedData original = SerializedData.from(val1, null);
        SerializedData data = original.copy(original.toJsonString().replace("Pojo1", "Pojo2").getBytes(), original.getType(), meta);
        System.out.println(data.toJsonString());

        Assertions.assertEquals(val2, data.deserialize(null));
    }

    @Test
    public void objectShouldBeDeserializableEvenWithLessProperties() {
        Pojo1 val1 = new Pojo1("42", 42, JType.TYPE_1);
        Pojo2 val2 = new Pojo2("42", 42);
        HashMap<String, byte[]> meta = new HashMap<>();
        meta.put(SerializedData.META_JAVA_CLASS, Pojo1.class.getName().getBytes(StandardCharsets.UTF_8));
        SerializedData original = SerializedData.from(val2, null);
        SerializedData data = original.copy(original.toJsonString().replace("Pojo2", "Pojo1").getBytes(), original.getType(), meta);
        System.out.println(data.toJsonString());

        Assertions.assertEquals(val1, data.deserialize(null));
    }

    @Test
    public void testPolymorphicDeserialize() {
        Pojo2 pojo = new Pojo2("42", 42);
        PojoWrapper val1 = new PojoWrapper(pojo, pojo);
        
        SerializedData data = SerializedData.from(val1, null);
        System.out.println(data.toJsonString());

        Assertions.assertEquals(val1, data.deserialize(null));
    }

//    @Test
//    public void ListOfSimpleObjectShouldBeSerializableDeserializable() {
//        Pojo1 pojoa = new Pojo1("42", 42, JType.TYPE_1);
//        Pojo1 pojob = new Pojo1("24", 24, JType.TYPE_2);
//        List<Pojo1> pojos = new ArrayList<>();
//        pojos.add(pojoa);
//        pojos.add(pojob);
//        SerializedData data = SerializedData.from(pojos, null);
//
//        System.out.println(data.toJsonString());
//
//        Assertions.assertEquals(pojos, data.deserialize(null));
//    }
}

enum JType {
    TYPE_1,
    TYPE_2
}

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "klass")
@JsonSubTypes({
        @JsonSubTypes.Type(value = Pojo1.class, name = "Pojo1"),
        @JsonSubTypes.Type(value = Pojo2.class, name = "Pojo2"),
        @JsonSubTypes.Type(value = Pojo3.class, name = "Pojo3"),
})
interface WithFoo {
    String getFoo();
}

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "klass")
@JsonSubTypes({
        @JsonSubTypes.Type(value = Pojo1.class, name = "Pojo1"),
        @JsonSubTypes.Type(value = Pojo2.class, name = "Pojo2")
})
abstract class Pojo {
    protected static String klass;
}

class Pojo1 extends Pojo implements WithFoo {
    private static final String klass = Pojo1.class.getSimpleName();
    private final String foo;
    private final int bar;
    private final JType type;

    @JsonCreator
    public Pojo1(
            @JsonProperty("foo") String foo,
            @JsonProperty("bar") int bar,
            @JsonProperty("type") JType type
    ) {
        this.foo = foo;
        this.bar = bar;
        this.type = type == null ? JType.TYPE_1 : type;
    }

    @JsonProperty("foo")
    public String getFoo() {
        return foo;
    }

    @JsonProperty("bar")
    public int getBar() {
        return bar;
    }

    @JsonProperty("type")
    public JType getType() {
        return type;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof Pojo1 that)) return false;
        return foo.equals(that.foo) && bar == that.bar && type.equals(that.type);
    }

    @Override
    public int hashCode() {
        int result = foo.hashCode();
        result = 31 * result + bar;
        result = 31 * result + type.hashCode();
        return result;
    }
}

class Pojo2 extends Pojo implements WithFoo {
    private static final String klass = Pojo2.class.getSimpleName();
    private final String foo;
    private final int bar;

    @JsonCreator
    public Pojo2(
            @JsonProperty("foo") String foo,
            @JsonProperty("bar") int bar
    ) {
        this.foo = foo;
        this.bar = bar;
    }

    @JsonProperty("foo")
    public String getFoo() {
        return foo;
    }

    @JsonProperty("bar")
    public int getBar() {
        return bar;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof Pojo2 that)) return false;
        return foo.equals(that.foo) && bar == that.bar;
    }

    @Override
    public int hashCode() {
        int result = foo.hashCode();
        result = 31 * result + bar;
        return result;
    }
}

record Pojo3(String foo, int bar) implements WithFoo {
    private static final String klass = "Pojo3";

    @Override
    public String getFoo() {
        return foo;
    }
}

record PojoWrapper(Pojo obj1, WithFoo obj2) {
}

File: infinitic-common/src/test/kotlin/io/infinitic/serDe/kotlin/SerDeTests.kt
package io.infinitic.serDe.kotlin

import io.infinitic.common.fixtures.TestFactory
import io.infinitic.common.serDe.SerializedData
import io.infinitic.common.workflows.data.steps.Step
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import kotlinx.serialization.Polymorphic
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.serializer
import java.lang.reflect.Method

class SerDeTests : StringSpec(
    {
      "Primitive (ByteArray) should be serializable / deserializable" {
        val bytes1: ByteArray = TestFactory.random()
        val bytes2 = SerializedData.from(bytes1).deserialize(null) as ByteArray

        bytes1.contentEquals(bytes2) shouldBe true
      }

      "Primitive (Short) should be serializable / deserializable" {
        val val1: Short = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Primitive (Int) should be serializable / deserializable" {
        val val1: Int = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Primitive (Long) should be serializable / deserializable" {
        val val1: Long = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Primitive (Float) should be serializable / deserializable" {
        val val1: Float = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Primitive (Double) should be serializable / deserializable" {
        val val1: Double = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Primitive (Boolean) should be serializable / deserializable" {
        val val1: Boolean = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Primitive (Char) should be serializable / deserializable" {
        val val1: Char = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "String should be serializable / deserializable" {
        val val1: String = TestFactory.random()
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "List Of primitives should be serializable / deserializable" {
        val val1 = listOf(42F, true, "!@#%")
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Null (Obj) should be serializable / deserializable" {
        val obj1: Obj1? = null
        val obj2 = SerializedData.from(obj1).deserialize(null)

        obj2 shouldBe obj1
      }

      "Simple Object should be serializable / deserializable" {
        val val1 = Obj1("42", 42, Type.TYPE_1)
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Object should be deserializable with more properties" {
        val val1 = Obj1("42", 42, Type.TYPE_1)
        val data = SerializedData.from(val1).copy(
            meta = mapOf(
                SerializedData.META_JAVA_CLASS to
                    Obj2::class.java.name.toByteArray(charset = Charsets.UTF_8),
            ),
        )
        val val2 = data.deserialize(null) as Obj2

        val2.foo shouldBe val1.foo
        val2.bar shouldBe val2.bar
      }

      "Object should be deserializable with less properties using default values" {
        val val1 = Obj2("42", 42)
        val data = SerializedData.from(val1).copy(
            meta = mapOf(
                SerializedData.META_JAVA_CLASS to
                    Obj1::class.java.name.toByteArray(charset = Charsets.UTF_8),
            ),
        )
        val val2 = data.deserialize(null) as Obj1

        val2.foo shouldBe val1.foo
        val2.bar shouldBe val1.bar
        val2.type shouldBe Type.TYPE_1
      }

      "Object containing set of sealed should be serializable / deserializable" {
        json = Json {
          classDiscriminator = "klass"
          serializersModule = SerializersModule {
            polymorphic(WithFoo::class, Obj1::class, Obj1.serializer())
            polymorphic(WithFoo::class, Obj2::class, Obj2.serializer())
          }
        }
        val obj = Obj1("42", 42, Type.TYPE_2)
        val val1 = Objs(setOf(obj), setOf(obj))
        val val2 = SerializedData.from(val1).deserialize(null)

        val2 shouldBe val1
      }

      "Step.Id should be serializable / deserializable" {
        val step1 = TestFactory.random<Step.Id>()
        val step2 = SerializedData.from(step1).deserialize(null)

        step2 shouldBe step1
      }

      "Step.Or should be serializable / deserializable" {
        val step1 = TestFactory.random<Step.Or>()
        val step2 = SerializedData.from(step1).deserialize(null)

        step2 shouldBe step1
      }

      "Step.And should be serializable / deserializable" {
        val step1 = TestFactory.random<Step.And>()
        val step2 = SerializedData.from(step1).deserialize(null)

        step2 shouldBe step1
      }

      "Comparing SerializedData" {
        val step = TestFactory.random<Step>()
        val s1 = SerializedData.from(step)
        val s2 = SerializedData.from(step)

        s1 shouldBe s2
      }

//      "List of simple Objects should be serializable / deserializable" {
//        val val1 = listOf(Obj1("42", 42, Type.TYPE_1), Obj1("24", 24, Type.TYPE_2))
//        val ser = SerializedData.from(val1)
//        println(ser.toJsonString())
//        val val2 = ser.deserialize(null)
//
//        val2 shouldBe val1
//      }
    },
)

fun main() {
  val json = Json {
    classDiscriminator = "#klass"

    serializersModule = SerializersModule {
      polymorphic(WithFoo::class, Obj1::class, Obj1.serializer())
      polymorphic(WithFoo::class, Obj2::class, Obj2.serializer())
    }
  }

  val method: Method = Obj1::class.java.getMethod("test", List::class.java)
  val type = method.parameters[0].parameterizedType
  val serializer = serializer(type)

  val o: List<Map<Int, WithFoo>> = listOf(mapOf(1 to Obj1("a", 2, Type.TYPE_1)))

  val oStr = json.encodeToString(serializer, o)

  println("oStr = $oStr")

  val o2: Any = json.decodeFromString(serializer, oStr)

  println("o2 = $o2")
  println("o = o2 : ${o == o2}")
}

@Serializable
private sealed class Obj

private enum class Type {
  TYPE_1,
  TYPE_2
}

@Serializable
@SerialName("Obj1")
private data class Obj1(
  override val foo: String,
  val bar: Int,
  val type: Type = Type.TYPE_1
) : Obj(), WithFoo {

  fun test(pojo: List<Map<Int, WithFoo>>) {
    println("from test(): $pojo")
  }

  fun test2(pojo: Obj) {
    println("from test2(): $pojo")
  }
}

@Serializable
@SerialName("Obj2")
private data class Obj2(override val foo: String, val bar: Int) : Obj(), WithFoo

@Serializable
private data class Objs(val objs: Set<Obj>, val others: Set<WithFoo>)

@Polymorphic
private sealed interface WithFoo {
  val foo: String
}
